### PR TITLE
TreeView component

### PR DESCRIPTION
**Pull request by kfaln**
This should be considered the first version of a work in progress.

The component needs some modification when the first (detail) view is to be implemented in the new Library, mainly the following.
- Async handling when fetching children for nodes.
- View routing/rendering when clicking on nodes (which view to display).

The TreeView currently handles static data, has working expand/collapse logic and styling.

### CODE REVIEW

File: src/components/TreeView/index.tsx [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        });
+
+        const newTreeData = [...treeData];
+        newTreeData.splice(collapsingNodeIndex + 1, childCount);
+
+        setTreeData(newTreeData);
+    };
+
+    const expandNode = (node: NodeData): void => {
+        if (!node.getChildren || node.isExpanded) {
```
Comment (Cnordbo) [inline src/components/TreeView/index.tsx]: is this ment to be `!node.hasChildren` ?

File: src/components/TreeView/index.tsx [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ * @param id Unique node identifier across all nodes (number or string).
+ * @param name Node display name.
+ * @param hasChildren Indicates node has children and can be expanded.
+ * @param getChildren Function to load children.
+ * @param onClick Function to load details view.
+ * @param parentId Internal: reference to parent node (handled by component).
+ * @param isExpanded Internal: indicates node is currently expanded (handled by component).
+ * @param children Internal: reference to child nodes (handled by component).
+ */
+export interface NodeData {
```
Comment (Cnordbo) [inline src/components/TreeView/index.tsx]: Is this an un-needed export?

File: src/components/TreeView/index.tsx [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    id: number | string;
+    name: string;
+    hasChildren: boolean;
+    getChildren: (() => NodeData[]) | null;
```
Comment (Cnordbo) [inline src/components/TreeView/index.tsx]: getChildren could probably always return a empty list, and should always be configured on the nodes.

File: src/components/TreeView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    onClick: (() => void) | null;
```
Comment (Cnordbo) [inline src/components/TreeView/index.tsx]: Optional property? 
`onClick?: (() => void)`

File: src/components/TreeView/index.tsx [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+    const getNodeLink = (node: NodeData): JSX.Element => {
+        const hasOnClickEvent = node.onClick;
+
+        return (
+            <NodeName 
+                hasChildren={node.hasChildren} 
+                isExpanded={node.isExpanded}>
+                {
+                    hasOnClickEvent && (
```
Comment (Cnordbo) [inline src/components/TreeView/index.tsx]: `node.onClick && ...`
No need for the extra variable,

File: src/components/TreeView/index.tsx [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+                        <NodeLink
+                            isExpanded={node.isExpanded} 
+                            onClick={(): void => {
```
Comment (Cnordbo) [inline src/components/TreeView/index.tsx]: Existence checking is already done before loading this code. 
Can be replaced with: `onClick={node.onClick}`

File: src/components/TreeView/index.tsx (hunk 1 above)
Comment (kfaln) [reply src/components/TreeView/index.tsx]: `hasChildren` was just a convenience prop, basically meaning the same as `getChildren` being defined/not null. `hasChildren` is removed/replaced with just checking `getChildren` (also moved elsewhere).

File: src/components/TreeView/index.tsx (hunk 2 above)
Comment (kfaln) [reply src/components/TreeView/index.tsx]: Parent components will need to send in data as a prop. The idea is that `NodeData` will be the data interface. However, it contains some props the parent won't need, so it's now split in two and only one of them is exported (`TreeViewNode`).

File: src/components/TreeView/index.tsx (hunk 3 above)
Comment (kfaln) [reply src/components/TreeView/index.tsx]: The parent container will need to supply functions for fetching children, if there are any. It's now changed to optional.

File: src/components/TreeView/index.tsx (hunk 4 above)
Comment (kfaln) [reply src/components/TreeView/index.tsx]: Yes.

File: src/components/TreeView/index.tsx (hunk 5 above)
Comment (kfaln) [reply src/components/TreeView/index.tsx]: Yes.

File: src/components/TreeView/index.tsx (hunk 6 above)
Comment (kfaln) [reply src/components/TreeView/index.tsx]: Yes.

**Pull request merged**